src/UhdmAst.cpp

```diff
@@ -1363,36 +1347,53 @@ AstNode* process_typespec(vpiHandle obj_h, UhdmShared& shared) {
     switch (objectType) {
     case vpiBitTypespec:
     case vpiLogicTypespec: {
-        return getDType(fl, obj_h, shared);
-    }
-    case vpiIntTypespec: {
-        auto* name = vpi_get_str(vpiName, obj_h);
-        if (name == nullptr) {
-            auto* dtypep = new AstBasicDType(fl, AstBasicDTypeKwd::INT);
-            return dtypep;
+        AstNodeDType* dtypep = nullptr;
+        if (auto elem_typespec_h = vpi_handle(vpiElemTypespec, obj_h)) {
+            dtypep = getDType(fl, elem_typespec_h, shared);
+            if (!dtypep) v3error("Unable to handle vpiElemTypespec node");
+        } else {
+            AstBasicDTypeKwd keyword = get_kwd_for_type(objectType);
+            dtypep = new AstBasicDType(fl, keyword);
         }
-        return new AstParseRef(fl, VParseRefExp::en::PX_TEXT, name, nullptr, nullptr);
-    }
-    case vpiStringTypespec: {
-        auto* dtypep = new AstBasicDType(fl, AstBasicDTypeKwd::STRING);
-        return dtypep;
-    }
-    case vpiChandleTypespec: {
-        auto* dtypep = new AstBasicDType(fl, AstBasicDTypeKwd::CHANDLE);
+        dtypep = applyPackedRanges(fl, obj_h, dtypep, shared);
         return dtypep;
     }
-    case vpiIntegerTypespec: {
-        AstNode* constNode = get_value_as_node(obj_h);
-        if (constNode == nullptr) {
-            v3info("Valueless typepec, returning dtype");
-            auto* dtypep = new AstBasicDType(fl, AstBasicDTypeKwd::INTEGER);
-            return dtypep;
-        }
-        return constNode;
+    case vpiIntTypespec:
+    case vpiLongIntTypespec:
+    case vpiIntegerTypespec:
+    case vpiByteTypespec:
+    case vpiRealTypespec:
+    case vpiStringTypespec:
+    case vpiChandleTypespec:
+    case vpiTimeTypespec: {
+        AstBasicDTypeKwd keyword = get_kwd_for_type(objectType);
+        return new AstBasicDType(fl, keyword);
     }
     case vpiVoidTypespec: {
         return new AstVoidDType(fl);
     }
+    case vpiArrayTypespec: {
+        auto elem_typespec_h = vpi_handle(vpiElemTypespec, obj_h);
+        AstNodeDType* dtypep = getDType(fl, elem_typespec_h, shared);
+        return applyUnpackedRanges(fl, obj_h, dtypep, shared);
+    }
+    case vpiPackedArrayTypespec: {
+        AstNodeDType* dtypep = nullptr;
+        auto elem_typespec_h = vpi_handle(vpiElemTypespec, obj_h);
+        if (elem_typespec_h) {
+            dtypep = getDType(fl, elem_typespec_h, shared);
+        } else {
+            UINFO(7, "No elem_typespec found in vpiPackedArrayTypespec, trying IndexTypespec"
+                         << std::endl);
+            auto index_typespec_h = vpi_handle(vpiIndexTypespec, obj_h);
+            if (index_typespec_h) {
+                dtypep = getDType(fl, index_typespec_h, shared);
+            } else {
+                v3error("\t! Failed to get typespec handle for PackedArrayTypespec");
```

Comment:
No need for '\t' in this error (I know it's probably copied, but this is a good opportunity to clean this up).